File: src/bst/binary_search_tree.cpp
#include <iostream>
using namespace std;

class Node
{
public:
    int key;
    Node *left;
    Node *right;

    Node(int key)
    {
        this->key = key;
        this->left = this->right = NULL;
    }
};

class BST
{
private:
    Node *root;

    Node *insert(Node *node, int key)
    {
        if (node == NULL)
        {
            return new Node(key);
        }
        if (key <= node->key)
        {
            node->left = insert(node->left, key);
        }
        else
        {
            node->right = insert(node->right, key);
        }
        return node;
    }

    Node *search(Node *node, int key)
    {
        if (node == NULL || node->key == key)
        {
            return node;
        }
        if (key < node->key)
        {
            return search(node->left, key);
        }
        return search(node->right, key);
    }

    void inOrder(Node *node)
    {
        if (node == NULL)
        {
            return;
        }
        inOrder(node->left);
        cout << node->key << " ";
        inOrder(node->right);
    }

    void preOrder(Node *node)
    {
        if (node == NULL)
        {
            return;
        }
        cout << node->key << " ";
        preOrder(node->left);
        preOrder(node->right);
    }

    void postOrder(Node *node)
    {
        if (node == NULL)
        {
            return;
        }
        postOrder(node->left);
        postOrder(node->right);
        cout << node->key << " ";
    }

public:
    BST()
    {
        this->root = NULL;
    }

    void insert(int key)
    {
        root = insert(root, key);
    }

    Node *search(int key)
    {
        return search(root, key);
    }

    void inOrder()
    {
        cout << "The inOrder traversal is: ";
        inOrder(root);
        cout << "\n";
    }

    void preOrder()
    {
        cout << "The preOrder traversal is: ";
        preOrder(root);
        cout << "\n";
    }

    void postOrder()
    {
        cout << "The postOrder traversal is: ";
        postOrder(root);
        cout << "\n";
    }
};

static void search(BST bst, int key)
{
    if (bst.search(key) != NULL)
    {
        cout << key << " found\n";
    }
    else
    {
        cout << key << " not found\n";
    }
}

int main()
{
    BST bst;
    bst.insert(10);
    bst.insert(15);
    bst.insert(5);
    bst.insert(8);
    bst.insert(18);
    bst.insert(12);
    bst.insert(10);

    bst.preOrder();
    bst.inOrder();
    bst.postOrder();

    search(bst, 12);
    search(bst, 9);
}

File: src/bst/binary_search_tree.py
class Node:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None


class BST:
    def __init__(self):
        self.root = None

    def insert(self, key):
        self.root = self.__insert__(self.root, key)

    def __insert__(self, node, key):
        if node is None:
            return Node(key)

        if key <= node.key:
            node.left = self.__insert__(node.left, key)

        else:
            node.right = self.__insert__(node.right, key)

        return node

    def search(self, key):
        return self.__search__(self.root, key)

    def __search__(self, node, key):
        if node is None or node.key == key:
            return node

        if key <= node.key:
            return self.__search__(node.left, key)

        return self.__search__(node.right, key)

    def inOrder(self):
        print("The inOrder traversal is: ", end=" ")
        self.__inOrder__(self.root)
        print()

    def __inOrder__(self, node):
        if node == None:
            return

        self.__inOrder__(node.left)
        print(node.key, end=" ")
        self.__inOrder__(node.right)

    def preOrder(self):
        print("The preOrder traversal is: ", end=" ")
        self.__preOrder__(self.root)
        print()

    def __preOrder__(self, node):
        if node == None:
            return

        print(node.key, end=" ")
        self.__preOrder__(node.left)
        self.__preOrder__(node.right)

    def postOrder(self):
        print("The preOrder traversal is: ", end=" ")
        self.__postOrder__(self.root)
        print()

    def __postOrder__(self, node):
        if node == None:
            return

        self.__postOrder__(node.left)
        self.__postOrder__(node.right)
        print(node.key, end=" ")


def search(bst, key):
    if bst.search(key) != None:
        print(key, "found")
    else:
        print(key, "not found")


bst = BST()
bst.insert(10)
bst.insert(15)
bst.insert(5)
bst.insert(8)
bst.insert(18)
bst.insert(12)
bst.insert(10)

bst.preOrder()
bst.inOrder()
bst.postOrder()

search(bst, 12)
search(bst, 9)
